apps/evm_family/evm_txn_helpers.c

```diff
@@ -174,56 +168,49 @@ static uint64_t get_decode_length(const uint8_t *seq,
   return item_bytes_len;
 }
 
-static PAYLOAD_STATUS evm_decode_txn_payload(
-    const evm_sign_txn_initiate_request_t *init_info,
-    const evm_unsigned_txn *utxn_ptr,
-    const erc20_contracts_t **contract) {
-  PAYLOAD_STATUS result = PAYLOAD_ABSENT;
-  evm_is_token_whitelisted = false;
-
-  if (utxn_ptr->payload_size > 0) {
-    if ((utxn_ptr->payload_size >= 4) &&
-        (U32_READ_BE_ARRAY(utxn_ptr->payload) == TRANSFER_FUNC_SIGNATURE) &&
-        (init_info->is_token_transfer) &&
-        (init_info->chain_id == ETHEREUM_MAINNET_CHAIN)) {
-      for (int16_t i = 0; i < g_evm_app->whitelist_count; i++) {
-        if (strncmp(init_info->token_symbol,
-                    g_evm_app->whitelisted_contracts[i].symbol,
-                    ETHEREUM_TOKEN_SYMBOL_LENGTH) == 0) {
-          evm_is_token_whitelisted = true;
-          result = (memcmp(utxn_ptr->to_address,
-                           g_evm_app->whitelisted_contracts[i].address,
-                           EVM_ADDRESS_LENGTH) == 0)
-                       ? PAYLOAD_WHITELISTED
-                       : PAYLOAD_CONTRACT_INVALID;
-
-          if (PAYLOAD_WHITELISTED == result) {
-            *contract = &g_evm_app->whitelisted_contracts[i];
-          }
-          break;
-        }
-      }
-    }
-    if (!evm_is_token_whitelisted) {
-      // TODO: detecting clear sign or blind sign transaction using the function
-      // tag value along with function signature
-      result =
-          (ETH_ExtractArguments(utxn_ptr->payload, utxn_ptr->payload_size) ==
-           ETH_UTXN_ABI_DECODE_OK)
-              ? PAYLOAD_WHITELISTED
-              : PAYLOAD_SIGNATURE_NOT_WHITELISTED;
+static EVM_TRANSACTION_TYPE evm_decode_transaction_type(
+    evm_txn_context_t *txn_context) {
+  if (0 == txn_context->transaction_info.data_size) {
+    return EVM_TXN_NO_DATA;
+  }
+
+  if (EVM_FUNC_SIGNATURE_LENGTH > txn_context->transaction_info.data_size) {
+    // function signature should be of 4-bytes
+    return EVM_TXN_INVALID_DATA;
+  }
+
+  uint32_t function_tag = U32_READ_BE_ARRAY(txn_context->transaction_info.data);
+  if (EVM_transfer_TAG == function_tag &&
+      g_evm_app->is_token_whitelisted(txn_context->transaction_info.to_address,
+                                      &txn_context->contract)) {
+    return EVM_TXN_TRANSFER_FUNC;
+  }
+
+  if (EVM_swap_TAG == function_tag || EVM_uniswapV3Swap_TAG == function_tag ||
+      EVM_safeTransferFrom_TAG == function_tag ||
+      EVM_deposit_TAG == function_tag || EVM_transfer_TAG == function_tag) {
+    // decode the contract data for display
+    if (ETH_UTXN_ABI_DECODE_OK !=
+        ETH_ExtractArguments(txn_context->transaction_info.data,
+                             txn_context->transaction_info.data_size,
+                             &txn_context->display_node)) {
+      return EVM_TXN_INVALID_DATA;
```

Comment:
Please add a comment here for which condition would be an Invalid case.

Reply:
added